test/Spec.hs

```diff
@@ -450,6 +452,89 @@ statefulTests =
                     ["default"]
          `shouldReturn` Just "12"
 
+doctypeTests :: SpecWith LarcenyHspecState
+doctypeTests = do
+  describe "doctypes" $ do
+    it "should render blank doctypes" $ do
+      "<doctype />" `shouldRenderM` "<!DOCTYPE html>"
+    it "should render regular doctype" $ do
+      "<!DOCTYPE html>" `shouldRenderM` "<!DOCTYPE html>"
+    it "should render doctype in the correct place" $ do
+      "<!DOCTYPE html><html><p>Hello world</p></html>"
+      `shouldRenderM` "<!DOCTYPE html><html><p>Hello world</p></html>"
+
+conditionalTests :: SpecWith LarcenyHspecState
+conditionalTests = do
+  describe "conditionals" $ do
+    let template cond =
+          "<if condition=\"" <> cond <> "\">\
+          \  <then>It's true!</then>\
+          \  <else>It's false!</else>\
+          \</if>"
+    describe "true condition" $ do
+      it "should display the stuff within the `then` tag" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)]
+        template "True" `shouldRenderM` "It's true!"
+      it "should work with a blank in the attribute" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)
+                ,("userIsLoggedIn", textFill "True")]
+        template "${userIsLoggedIn}" `shouldRenderM` "It's true!"
+    describe "false condition" $ do
+      it "should display the stuff within the `else` tag" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)]
+        template "False" `shouldRenderM` "It's false!"
+      it "should display the stuff within the `else` tag" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)
+                ,("userIsLoggedIn", textFill "False")]
+        template "${userIsLoggedIn}" `shouldRenderM` "It's false!"
+  describe "exists" $ do
+    let template =
+          "<if exists=\"${existing}\">\
+          \  <then>It <existing />!</then>\
+          \  <else>It doesn't exist!</else>\
+          \</if>"
+    describe "the fill exists" $ do
+      it "should display the stuff within the `then` tag" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)
+                ,("existing", textFill "exists")]
+        template `shouldRenderM` "It exists!"
+    describe "the fill doesn't exist" $ do
+      it "should display the stuff within the `else` tag" $ do
+        hLarcenyState.lSubs .=
+           subs [("if", ifFill)]
+        template `shouldRenderM` "It doesn't exist!"
```

Comment:
also test "False" and "" fills for `existing`

Reply:
ummmmm how do you feel about the specified behavior of "exists" being that if it's a blank string then it doesn't exist? test:

```(Haskell)
describe "the fill is an empty string" $
      it "should display the stuff within the `else` tag" $ do
        hLarcenyState.lSubs .=
           subs [("if", ifFill)
                ,("existing", textFill "")]
        template `shouldRenderM` "It doesn't exist!"
```

Reply:
as a first approximation I mean

Reply:
👍